pgvectorscale/src/access_method/labels.rs

```diff
@@ -0,0 +1,271 @@
+    fn from(labels: &Vec<Label>) -> Self {
+        let mut labels = labels.clone();
+        labels.sort_unstable();
+        labels.dedup();
+
+        Self { labels }
+    }
+}
+
+impl From<&ArchivedLabelSet> for LabelSet {
+    fn from(set: &ArchivedLabelSet) -> Self {
+        Self {
+            labels: set.labels.to_vec(),
+        }
+    }
+}
+
+impl From<ArchivedLabelSet> for LabelSet {
+    fn from(set: ArchivedLabelSet) -> Self {
+        Self {
+            labels: set.labels.to_vec(),
+        }
+    }
+}
+
+impl From<Label> for LabelSet {
+    fn from(label: Label) -> Self {
+        Self {
+            labels: vec![label],
+        }
+    }
+}
+
+impl FromIterator<Label> for LabelSet {
+    fn from_iter<T: IntoIterator<Item = Label>>(iter: T) -> Self {
+        let mut labels: Vec<Label> = iter.into_iter().collect();
+        labels.sort_unstable();
+        labels.dedup();
+
+        Self { labels }
+    }
+}
+
+/// Accessor trait to allow common logic between `LabelSet` and `ArchivedLabelSet`.
+pub trait LabelSetView {
+    fn labels(&self) -> &[Label];
+
+    /// Any labels in the labelset?
+    fn is_empty(&self) -> bool {
+        self.labels().is_empty()
+    }
+
+    /// Do the labelsets share any labels?
+    fn overlaps<T: LabelSetView>(&self, other: &T) -> bool {
+        let a = self.labels();
+        let b = other.labels();
+
+        debug_assert!(a.is_sorted());
+        debug_assert!(b.is_sorted());
+
+        let mut i = 0;
+        let mut j = 0;
+
+        while i < a.len() && j < b.len() {
+            match a[i].cmp(&b[j]) {
+                std::cmp::Ordering::Equal => return true,
+                std::cmp::Ordering::Less => i += 1,
+                std::cmp::Ordering::Greater => j += 1,
+            }
+        }
+        false
+    }
+
+    fn iter(&self) -> std::slice::Iter<Label> {
+        self.labels().iter()
+    }
+}
+
+impl LabelSetView for LabelSet {
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+impl LabelSetView for ArchivedLabelSet {
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+impl Debug for ArchivedLabelSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ArchivedLabelSet")
+            .field(&self.labels.to_vec())
+            .finish()
+    }
+}
+
+/// A labeled vector is a vector with an optional set of labels.
+#[derive(Debug)]
+pub struct LabeledVector {
+    vec: PgVector,
+    labels: Option<LabelSet>,
+}
+
+impl LabeledVector {
+    pub fn new(vec: PgVector, labels: Option<LabelSet>) -> Self {
+        Self { vec, labels }
+    }
+
+    pub unsafe fn from_datums(
+        values: *mut Datum,
+        isnull: *mut bool,
+        meta_page: &MetaPage,
+    ) -> Option<Self> {
+        let vec = PgVector::from_pg_parts(values, isnull, 0, meta_page, true, false)?;
+
+        let labels: Option<LabelSet> = if meta_page.has_labels() {
+            if *isnull.add(1) {
```

Comment:
is the add(1) correct here? The datum comes from the table layout not the index layout in some cases, right? what if the table column ordering is different than
```
            embedding vector(3),
            labels SMALLINT[],
```

what if we have

```
            labels SMALLINT[],
            embedding vector(3),
```

or 

some columns in-between

We should add test-cases for all of these.